cpp/src/plasma/client.cc

```diff
@@ -0,0 +1,593 @@
+#endif
+
+#include <assert.h>
+#include <fcntl.h>
+#include <netinet/in.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <strings.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/un.h>
+#include <unistd.h>
+
+#include "plasma/client.h"
+#include "plasma/common.h"
+#include "plasma/io.h"
+#include "plasma/plasma.h"
+#include "plasma/protocol.h"
+
+#include <algorithm>
+#include <thread>
+#include <vector>
+
+#include "plasma/fling.h"
+
+#define XXH_STATIC_LINKING_ONLY
+#include "thirdparty/xxhash.h"
+
+#define XXH64_DEFAULT_SEED 0
+
+// Number of threads used for memcopy and hash computations.
+constexpr int64_t kThreadPoolSize = 8;
+constexpr int64_t kBytesInMB = 1 << 20;
+static std::vector<std::thread> threadpool_(kThreadPoolSize);
+
+struct ClientMmapTableEntry {
+  /// The result of mmap for this file descriptor.
+  uint8_t* pointer;
+  /// The length of the memory-mapped file.
+  size_t length;
+  /// The number of objects in this memory-mapped file that are currently being
+  /// used by the client. When this count reaches zeros, we unmap the file.
+  int count;
+};
+
+struct ObjectInUseEntry {
+  /// A count of the number of times this client has called PlasmaClient::Create
+  /// or
+  /// PlasmaClient::Get on this object ID minus the number of calls to
+  /// PlasmaClient::Release.
+  /// When this count reaches zero, we remove the entry from the ObjectsInUse
+  /// and decrement a count in the relevant ClientMmapTableEntry.
+  int count;
+  /// Cached information to read the object.
+  PlasmaObject object;
+  /// A flag representing whether the object has been sealed.
+  bool is_sealed;
+};
+
+// If the file descriptor fd has been mmapped in this client process before,
+// return the pointer that was returned by mmap, otherwise mmap it and store the
+// pointer in a hash table.
+uint8_t* lookup_or_mmap(PlasmaClient* conn, int fd, int store_fd_val, int64_t map_size) {
+  auto entry = conn->mmap_table.find(store_fd_val);
+  if (entry != conn->mmap_table.end()) {
+    close(fd);
+    return entry->second->pointer;
+  } else {
+    uint8_t* result = reinterpret_cast<uint8_t*>(
+        mmap(NULL, map_size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0));
+    if (result == MAP_FAILED) { ARROW_LOG(FATAL) << "mmap failed"; }
+    close(fd);
+    ClientMmapTableEntry* entry = new ClientMmapTableEntry();
+    entry->pointer = result;
+    entry->length = map_size;
+    entry->count = 0;
+    conn->mmap_table[store_fd_val] = entry;
+    return result;
+  }
+}
+
+// Get a pointer to a file that we know has been memory mapped in this client
+// process before.
+uint8_t* lookup_mmapped_file(PlasmaClient* conn, int store_fd_val) {
+  auto entry = conn->mmap_table.find(store_fd_val);
+  ARROW_CHECK(entry != conn->mmap_table.end());
+  return entry->second->pointer;
+}
+
+void increment_object_count(
+    PlasmaClient* conn, ObjectID object_id, PlasmaObject* object, bool is_sealed) {
```

Comment:
I'm doing that everywhere in the codebase now (where applicable). Let me know if you spot an instance I forgot.